cpp/src/fil/common.cuh

```diff
@@ -173,6 +172,89 @@ struct predict_params : shmem_size_params {
   int num_blocks;
 };
 
+namespace dispatch {
+
+template <template <bool, leaf_algo_t, int> class Func, typename storage_type,
+          bool cols_in_shmem, leaf_algo_t leaf_algo, int n_items,
+          typename... Args>
+void dispatch_on_n_items(predict_params& params, Args... args) {
+  ASSERT(params.n_items <= 4, "internal error: n_items > 4");
```

Comment:
This assertion is checked for every value of `n_items`. It is definitely better to check it only once.

Besides, it doesn't catch the situation when `n_items <= 0`, something the previous version catches.